### PR TITLE
Fix unzipping 4GB+ zip files

### LINKED ISSUE #77159: DotNet 7 Regression when reading large Zip files
### Description

When trying to read `ZipArchiveEntry`s from an 8GB zip file that contains ~8K ~1MB files I start getting `A local file header is corrupt.` exception. This happens _only once I get passed the "~4GB" mark_ (i.e. those entries that are located within first 4GB are all read just fine, but all the rest are all unreadable).

After some research, I realized that very likely culprit is in `src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs`, likely [this PR](https://github.com/dotnet/runtime/pull/68106) ([commit](https://github.com/dotnet/runtime/commit/9a4cdce1a20da9a69de59de2d55cdc3df7ade9b7)). I traced the issue by stepping through in VS to this place:

https://github.com/dotnet/runtime/blob/10fc8ae9a2af0a0ced69de92bb0d7734ff35946b/src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs#L768-L773

`_offsetOfLocalHeader` is set to `0xffffffff` and is being passed directly into the reader which tries to read the data, which is obviously wrong. Instead it should be using the offset from the Zip64 extended info.

I have verified using multiple tools like 7zip that the zip file is completely valid. I have also repeatedly reproed the issue by producing various >4GB zip files in various tools - all with the same result: _entries that are beyond the 4GB offset_ are not readable and all return `A local file header is corrupt.` exception.

### Reproduction Steps

Just try calling `ZipArchiveEntry.Open()` on an entry in a zip archive which is located beyond the 4GB mark (in the archive file).

### Expected behavior

As it was in DotNet 6 (which is the last version I used where it worked). Should be able to read entries in a zip archive beyond the 4 GB mark.

### Actual behavior

None of the zip entries beyond the 4GB mark can be opened.

### Regression?

Yes, tried it on RC1, but looks like it was introduced at least in Preview 5, possibly earlier. And by the looks of it present in RC2 since the affected files have not changed for RC2.

### Known Workarounds

Use other library. (Not really a workaround though).

### Configuration

```
.NET SDK:
 Version:   7.0.100-rc.1.22431.12
 Commit:    f1cf61e1c0

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.17134
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\7.0.100-rc.1.22431.12\

Host:
  Version:      7.0.0-rc.1.22426.10
  Architecture: x64
  Commit:       06aceb7015

.NET SDKs installed:
  5.0.400-preview.21328.4 [C:\Program Files\dotnet\sdk]
  6.0.100-preview.5.21302.13 [C:\Program Files\dotnet\sdk]
  6.0.400 [C:\Program Files\dotnet\sdk]
  7.0.100-rc.1.22431.12 [C:\Program Files\dotnet\sdk]
```

### Other information

Hopefully either @danmoseley or @adamsitnik can take a look ASAP. This is a serious regression that prevents a whole lot of valid scenarios!

### PR DESCRIPTION
**Pull request by adamsitnik**
fixes #77159 without reverting #68106

I've tried to keep the changes as small as possible to make it easier to review and backport.

Explanation: for a file that is over 4 GB after zipping, some of the entries were reporting `extraField.Size == 8`, but `readUncompressedSize`, `readCompressedSize` and `readStartDiskNumber` equal `false` and `readLocalHeaderOffset` equal true. So what the code was doing, was reading first int64, not using it and exiting:

https://github.com/dotnet/runtime/blob/f7e6cf145e4b18f12675d7f0f0eb7d5216cd1a80/src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs#L202-L207

The fix is to allow advancing the stream (by reading from it) only when all 4 fields are provided (`extraField.Size == 28`) or when they are explicitly requested (the boolean flags set to true)

### CODE REVIEW

File: src/libraries/System.IO.Compression/tests/ZipArchive/zip_LargeFiles.cs
Comment (iSazonov) [inline src/libraries/System.IO.Compression/tests/ZipArchive/zip_LargeFiles.cs:34]: Maybe use CompressionLevel.Fastest or NoCompression to save time?

Comment (adamsitnik) [reply src/libraries/System.IO.Compression/tests/ZipArchive/zip_LargeFiles.cs:34]: > Maybe use CompressionLevel.Fastest or NoCompression to save time?

@iSazonov thank you, it's a good suggestion!

I've tried it and for `CompressionLevel.NoCompression` it takes 10 minutes to run the test, while for `CompressionLevel.Fastest` it's 12 minutes, almost identical to current implementation (`CompressionLevel.Optimal` is the default).

Since there is no big difference between 10 minutes and 12 and `CompressionLevel.Optimal` is closer to real life scenarios I am going to keep it as it is for now (I am also going to experiment with fewer, but larger files).

Comment (iSazonov) [reply src/libraries/System.IO.Compression/tests/ZipArchive/zip_LargeFiles.cs:34]: I guess `random.NextBytes` takes ~10%.
So CI disk is too slow. Maybe use a RAM disk if possible?

File: src/libraries/System.IO.Compression/tests/ZipArchive/zip_LargeFiles.cs
Comment (danmoseley) [inline src/libraries/System.IO.Compression/tests/ZipArchive/zip_LargeFiles.cs:14]: this probably needs the special xunit collection so it doesn't run in parallel with any other tests.

File: src/libraries/System.IO.Compression/tests/ZipArchive/zip_LargeFiles.cs
Comment (danmoseley) [inline src/libraries/System.IO.Compression/tests/ZipArchive/zip_LargeFiles.cs:25]: if you're not going to validate the content of the buffers, I guess you can just fill it once and reuse it for each file.

File: src/libraries/System.IO.Compression/tests/ZipArchive/zip_LargeFiles.cs
Comment (danmoseley) [inline src/libraries/System.IO.Compression/tests/ZipArchive/zip_LargeFiles.cs:44]: does'nt matter much, but I think File.Delete() succeeds if the file doesn't exist.

File: src/libraries/System.IO.Compression/tests/ZipArchive/zip_LargeFiles.cs
Comment (danmoseley) [reply src/libraries/System.IO.Compression/tests/ZipArchive/zip_LargeFiles.cs:14]: https://github.com/dotnet/runtime/issues/77012#issuecomment-1277992442

File: src/libraries/System.IO.Compression/tests/ZipArchive/zip_LargeFiles.cs
Comment (adamsitnik) [reply src/libraries/System.IO.Compression/tests/ZipArchive/zip_LargeFiles.cs]: @danmoseley you are right (I was not aware of that, but I've verified that)

Comment (danmoseley) [reply src/libraries/System.IO.Compression/tests/ZipArchive/zip_LargeFiles.cs]: similar to Directory.CreateDirectory does not need a check it doesn't exist.

**Comment by adamsitnik** [comment]
> Seems reasonable but I'd rather someone on the IO crew sign off on the change.

@Jozkee PTAL

**Comment by Jozkee** [comment]
Taking a look...

File: src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs
Comment (Jozkee) [inline src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs:214]: What does -1 is meant to accomplish/signal?

File: src/libraries/System.IO.Compression/tests/ZipArchive/zip_LargeFiles.cs
Comment (Jozkee) [inline src/libraries/System.IO.Compression/tests/ZipArchive/zip_LargeFiles.cs:17]: Technically, this is not 1Gb, but since you are creating 6 files, the test is still valid. 

I would rather create two tests, one that succeeds at the limit (say 3.999 Gb) without the fix and one that fails and then verify that both pass with your fix. 

Since this is urgent for a backport, we can defer that.

File: src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs
Comment (Jozkee) [inline src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs:219]: Should remaining fields be set to -1 if we return before attempting to read them?

File: src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs
Comment (Jozkee) [inline src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs:207]: Could this fail if only 2 or 3 fields are available for reading but the bool params are false?

i.e: what if extraField.size = 16 and readUncompressedSize = false and readCompressedSize = false?

File: src/libraries/System.IO.Compression/tests/ZipArchive/zip_LargeFiles.cs
Comment (adamsitnik) [reply src/libraries/System.IO.Compression/tests/ZipArchive/zip_LargeFiles.cs:17]: > Technically, this is not 1Gb, but since you are creating 6 files, the test is still valid.

Comment (adamsitnik) [reply src/libraries/System.IO.Compression/tests/ZipArchive/zip_LargeFiles.cs:17]: > Technically, this is not 1Gb, but since you are creating 6 files, the test is still valid.

You mean that $1000^3$ is not 1 GB but $1024^3$ is?

> I would rather create two tests, one that succeeds at the limit (say 3.999 Gb) without the fix and one that fails and then verify that both pass with your fix.

My typical workflow is to add test that fails first, then make it pass. We have plenty of tests that verify < 4 GB archives so I don't see the value in adding it (also it would require us to use even more disk space and take longer to execute). 

In the long term we could add more tests, especially such that verify that `decompress(compress(x)) == x` (which this test avoids, as it was taking 10 minutes to do so on my beefy PC).

File: src/libraries/System.IO.Compression/tests/ZipArchive/zip_LargeFiles.cs
Comment (adamsitnik) [reply src/libraries/System.IO.Compression/tests/ZipArchive/zip_LargeFiles.cs:20]: Good catch! I'll fix it.

File: src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs
Comment (adamsitnik) [reply src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs:214]: > What does -1 is meant to accomplish/signal?

"Undefined, don't use".

Initially I've used something like this:

```cs
long value64;

if (readUncompressedSize || readAllFields)
{
    value64 = reader.ReadInt64();
    
    if (readUncompressedSize)
    {
        zip64Block._uncompressedSize = value64;
    }
}
```

But the diff was more complex, so I decided to "simplify it" by having such one liner changes (`condition ? read : -1`).

FWIW the values are validated at the end:

https://github.com/dotnet/runtime/blob/50b65de0d6deaa45fd48e6991fedb1a029e9af28/src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs#L233-L236

The negative value is never used (I just needed something for the `? :` conditon).

File: src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs
Comment (adamsitnik) [reply src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs:207]: I don't think that this is possible. @danmoseley has added following comment:

https://github.com/dotnet/runtime/blob/50b65de0d6deaa45fd48e6991fedb1a029e9af28/src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs#L190-L197

My understanding is that we have two options:

1. Correct archives: the fields are provided when the directory record fields are set (the right booleans are set to true, the size is their total size)
2. "Slightly incorrect archives": the boolean fields are not set, but the provided size suggests that they are all present. Tests added in https://github.com/dotnet/runtime/pull/68106 use such approach.

File: src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs
Comment (Jozkee) [reply src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs:214]: > FWIW the values are validated at the end:

Those validation only occur if extraField.Size >= 28. Shouldn't we validate each time reader.ReadInt* is called?

File: src/libraries/System.IO.Compression/tests/ZipArchive/zip_LargeFiles.cs
Comment (Jozkee) [reply src/libraries/System.IO.Compression/tests/ZipArchive/zip_LargeFiles.cs:17]: > You mean that $1000^3$ is not 1 GB but $1024^3$ is?

Yes.

> We have plenty of tests that verify < 4 GB archives so I don't see the value in adding it (also it would require us to use even more disk space and take longer to execute).

I meant to verify on the edge cases of the bug. In this case that would be 4Gb - 1, 4Gb, and 4Gb + 1; rather than a 6Gb case. But yes, that would multiply test execution time which is already quite large with this new test.

File: src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs
Comment (stephentoub) [inline src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs:204]: If you're going to include issue/PR numbers, please include the full link.  We've already moved code between repos once, and the meaning of such numbers changes when we do.

File: src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs
Comment (stephentoub) [reply src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs:214]: > The negative value is never used (I just needed something for the ? : conditon).

There doesn't actually need to be an else.  As long as you initialize value64 once when it's declared, each of these conditions could be of the form:
```C#
if (readLocalHeaderOffset || readAllFields)
{
    value64 = reader.ReadInt64();
}
```
It might avoid some confusion.  You could also be more explicit about the advance-but-ignore case, if desired, e.g.
```C#
if (readLocalHeaderOffset)
{
    zip64Block._localHeaderOffset = reader.ReadInt64();
}
else
{
    _ = reader.ReadInt64(); // advance
}
```

That said, what you have seems ok.

File: src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs
Comment (danmoseley) [reply src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs:204]: My personal taste is to not include them in product code, and rely on comments, links in tests, and commit comments. But that's from seeing code bases full up with 20 years of bug numbers, and get harder to read for it. But I am not sure that is a consensus view 🙂

**Comment by adamsitnik** [comment]
@Jozkee @stephentoub @danmoseley ping

File: src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs
Comment (Jozkee) [inline src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs:253]: These validations only occur if extraField.Size >= 28 since you are now returning if there's less than that. 

Shouldn't we validate when `reader.ReadInt*` is called and/or if read of the field was requested?

File: src/libraries/System.IO.Compression/tests/ZipArchive/zip_LargeFiles.cs
Comment (stephentoub) [inline src/libraries/System.IO.Compression/tests/ZipArchive/zip_LargeFiles.cs:17]: Is there a reasonable chance this will OOM?  If so, consider catching that and skipping the test, rather than failing it.

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.IO.Compression/tests/ZipArchive/zip_LargeFiles.cs
Comment (adamsitnik) [reply src/libraries/System.IO.Compression/tests/ZipArchive/zip_LargeFiles.cs:17]: > Is there a reasonable chance this will OOM? 

It should be executed only in Outerloop runs for the 64 bit "fast runtimes" and not in parallel with other tests so I don't think it's needed (at least for now).

https://github.com/dotnet/runtime/blob/4979dff15273a2c33238a5d393f4895e9c94e538/src/libraries/System.IO.Compression/tests/ZipArchive/zip_LargeFiles.cs#L13

https://github.com/dotnet/runtime/blob/4979dff15273a2c33238a5d393f4895e9c94e538/src/libraries/System.IO.Compression/tests/ZipArchive/zip_LargeFiles.cs#L10

https://github.com/dotnet/runtime/blob/98a89c9d7444f219c7b1aeb3796525d35f04d540/src/libraries/Common/tests/TestUtilities/System/PlatformDetection.cs#L105-L106

File: src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs
Comment (adamsitnik) [reply src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs:253]: This is how it always was, I would prefer not to do it now as I intend to backport it to 7.0 and trying to make it as defensive/safe as possible.

File: src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs
Comment (carlossanlop) [inline src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs:213]: For my own education: is there a perf improvement when using `_ = ReadInt64();` to discard the value, compared to just not using `_`? Or is the idea here to explicitly signal that we are ignoring the value on purpose?

File: src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs
Comment (carlossanlop) [reply src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs:229]: If the idea is to keep this PR minimal for backporting, this should probably be done later.

An alternative for this could be to set all those values to `-1` initially, then only the ones that are read before returning early will acquire a different value than -1.

**Comment by adamsitnik** [comment]
@danmoseley @Jozkee @stephentoub thank you for the reviews!

**Review by carlossanlop — COMMENTED** [review]
Assuming the CI is green, this LGTM. I'd also wait for a sign-off from @Jozkee.

Edit: Nevermind. This was merged while I was still adding a review. :)

File: src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs
Comment (Jozkee) [reply src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs:253]: The reason I was asking is because it wasn't always like that. It was changed to how it is now on https://github.com/dotnet/runtime/pull/68106, which is the PR that introduced the regression.

File: src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs
Comment (danmoseley) [reply src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs:213]: It shouldn't have a runtime effect, it's self documenting, both for humans and also for linters/analyzers that might otherwise flag an unused method result.

**Pull request merged**
